### PR TITLE
[Refactor] Python frontend refactor: build_call part, support format print and fstring

### PR DESCRIPTION
**Pull request by Leonz5288**
Related issue = #3038, #3233

### CODE REVIEW

File: python/taichi/lang/ir_builder.py [hunk 1]
```diff
@@ -146,7 +161,31 @@ def build_Call(ctx, node):
             else:
                 args.append(arg.ptr)
         keywords = dict(ChainMap(*[keyword.ptr for keyword in node.keywords]))
-        node.ptr = node.func.ptr(*args, **keywords)
+
+        if isinstance(node.func, ast.Attribute):
+            attr_name = node.func.attr
+            if attr_name == 'format':
+                args.insert(0, node.func.value.ptr)
+                node.ptr = ti.ti_format(*args, **keywords)
```
Comment (lin-hitonami) [inline python/taichi/lang/ir_builder.py:169]: add an else here?

File: python/taichi/lang/ir_builder.py [hunk 2]
```diff
@@ -146,7 +161,31 @@ def build_Call(ctx, node):
             else:
                 args.append(arg.ptr)
         keywords = dict(ChainMap(*[keyword.ptr for keyword in node.keywords]))
-        node.ptr = node.func.ptr(*args, **keywords)
+
+        if isinstance(node.func, ast.Attribute):
+            attr_name = node.func.attr
+            if attr_name == 'format':
```
Comment (lin-hitonami) [inline python/taichi/lang/ir_builder.py:167]: Should we check if node.func.value.ptr is str?

Comment (Leonz5288) [reply python/taichi/lang/ir_builder.py:167]: Make sense

File: python/taichi/lang/ir_builder.py (hunk 1 above)
Comment (Leonz5288) [reply python/taichi/lang/ir_builder.py:169]: ok

File: python/taichi/lang/ir_builder.py [hunk 3]
```diff
@@ -146,7 +161,31 @@ def build_Call(ctx, node):
...
+                args.insert(0, node.func.value.ptr)
+                node.ptr = ti.ti_format(*args, **keywords)
+        if isinstance(node.func, ast.Name):
```
Comment (lin-hitonami) [inline python/taichi/lang/ir_builder.py:170]: if->elif, and add an else to this one too

File: python/taichi/lang/ir_builder.py (hunk 2 above)
Comment (Leonz5288) [reply python/taichi/lang/ir_builder.py:167]: But what else could it be?

Comment (lin-hitonami) [reply python/taichi/lang/ir_builder.py:169]: Like an instance of a class with a method named "format"?

File: python/taichi/lang/ir_builder.py (hunk 3 above)
Comment (lin-hitonami) [reply python/taichi/lang/ir_builder.py:175]: is this one resolved?

Comment (Leonz5288) [reply python/taichi/lang/ir_builder.py:175]: resolved now

File: python/taichi/lang/ir_builder.py [hunk 4]
```diff
@@ -146,7 +163,36 @@ def build_Call(ctx, node):
             else:
                 args.append(arg.ptr)
         keywords = dict(ChainMap(*[keyword.ptr for keyword in node.keywords]))
-        node.ptr = node.func.ptr(*args, **keywords)
+
+        if isinstance(node.func, ast.Attribute):
+            attr_name = node.func.attr
+            if attr_name == 'format':
+                assert isinstance(node.func.value.ptr, str)
```
Comment (lin-hitonami) [inline python/taichi/lang/ir_builder.py:170]: this should be a part of if condition?

Comment (Leonz5288) [reply python/taichi/lang/ir_builder.py:170]: OK

**Review by lin-hitonami — APPROVED** [review]
LGTM! BTW, would you mind adding a test in tests/python/test_ast_refactor.py?

**Comment by Leonz5288** [comment]
> LGTM! BTW, would you mind adding a test in tests/python/test_ast_refactor.py?

No problem!

**Pull request merged**